examples/ex23.cpp

```diff
@@ -0,0 +1,988 @@
+         break;
+      }
+      case beam:
+      {
+         // T_10 mode
+         if (dim == 3)
+         {
+            double k10 = sqrt(k * k - M_PI * M_PI);
+            E[1] = -zi * k / M_PI * sin(M_PI*x(2))*exp(zi * k10 * x(0));
+         }
+         else if (dim == 2)
+         {
+            E[1] = -zi * k / M_PI * exp(zi * k * x(0));
+         }
+         break;
+      }
+      default:
+         break;
+   }
+}
+
+void E_exact_Re(const Vector &x, Vector &E)
+{
+   std::vector<std::complex<double>> Eval(E.Size());
+   maxwell_solution(x, Eval);
+   for (int i = 0; i < dim; ++i)
+   {
+      E[i] = Eval[i].real();
+   }
+}
+
+void E_exact_Im(const Vector &x, Vector &E)
+{
+   std::vector<std::complex<double>> Eval(E.Size());
+   maxwell_solution(x, Eval);
+   for (int i = 0; i < dim; ++i)
+   {
+      E[i] = Eval[i].imag();
+   }
+}
+
+void E_bdr_data_Re(const Vector &x, Vector &E)
+{
+   E = 0.0;
+   bool in_pml = false;
+
+   for (int i = 0; i < dim; ++i)
+   {
+      if (x(i) - comp_domain_bdr(i, 0) < 0.0 ||
+          x(i) - comp_domain_bdr(i, 1) > 0.0)
+      {
+         in_pml = true;
+         break;
+      }
+   }
+   if (!in_pml)
+   {
+      std::vector<std::complex<double>> Eval(E.Size());
+      maxwell_solution(x, Eval);
+      for (int i = 0; i < dim; ++i)
+      {
+         E[i] = Eval[i].real();
+      }
+   }
+}
+
+void E_bdr_data_Im(const Vector &x, Vector &E)
+{
+   E = 0.0;
+   bool in_pml = false;
+
+   for (int i = 0; i < dim; ++i)
+   {
+      if (x(i) - comp_domain_bdr(i, 0) < 0.0 ||
+          x(i) - comp_domain_bdr(i, 1) > 0.0)
+      {
+         in_pml = true;
+         break;
+      }
+   }
+   if (!in_pml)
+   {
+      std::vector<std::complex<double>> Eval(E.Size());
+      maxwell_solution(x, Eval);
+      for (int i = 0; i < dim; ++i)
+      {
+         E[i] = Eval[i].imag();
+      }
+   }
+}
+
+void detJ_JT_J_inv_Re(const Vector &x, CartesianPML * pml ,  DenseMatrix &M)
+{
+   std::vector<std::complex<double>> dxs(dim);
+   complex<double> det(1.0, 0.0);
+   pml->StretchFunction(x, dxs);
+
+   for (int i = 0; i < dim; ++i)
+   {
+      det *= dxs[i];
+   }
+
+   M = 0.0;
+   for (int i = 0; i < dim; ++i)
+   {
+      M(i, i) = (det / pow(dxs[i], 2)).real();
+   }
+}
+
+void detJ_JT_J_inv_Im(const Vector &x, CartesianPML * pml,  DenseMatrix &M)
+{
+   std::vector<std::complex<double>> dxs(dim);
+   complex<double> det = 1.0;
+   pml->StretchFunction(x, dxs);
+
+   for (int i = 0; i < dim; ++i)
+   {
+      det *= dxs[i];
+   }
+
+   M = 0.0;
+   for (int i = 0; i < dim; ++i)
+   {
+      M(i, i) = (det / pow(dxs[i], 2)).imag();
+   }
+}
+
+void detJ_JT_J_inv_abs(const Vector &x, CartesianPML * pml,  DenseMatrix &M)
+{
+   std::vector<std::complex<double>> dxs(dim);
+   complex<double> det = 1.0;
+   pml->StretchFunction(x, dxs);
+
+   for (int i = 0; i < dim; ++i)
+   {
+      det *= dxs[i];
+   }
+
+   M = 0.0;
+   for (int i = 0; i < dim; ++i)
+   {
+      M(i, i) = abs(det / pow(dxs[i], 2));
```

Comment:
There is some ambiguity in the type used by abs(). Although this works, it may be clearer to use fabs or std::abs.

Reply:
fabs will not work here since the argument is complex<double>